crates/optimism/payload/src/builder.rs

```diff
@@ -541,3 +552,69 @@ where
 
     Ok(BuildOutcome::Better { payload, cached_reads })
 }
+
+/// Extracts the Holcene 1599 parameters from the encoded form:
+/// <https://github.com/ethereum-optimism/specs/blob/main/specs/protocol/holocene/exec-engine.md#eip1559params-encoding>
+pub fn decode_eip_1559_params(eip_1559_params: B64) -> (u32, u32) {
+    let elasticity: [u8; 4] = eip_1559_params.0[..4].try_into().unwrap();
+    let denominator: [u8; 4] = eip_1559_params.0[4..].try_into().unwrap();
+    println!("denominator: {:?}", denominator);
+    println!("elasticity: {:?}", elasticity);
+    (u32::from_be_bytes(elasticity), u32::from_be_bytes(denominator))
+}
+
+fn get_holocene_extra_data(
+    attributes: &OptimismPayloadBuilderAttributes,
+    default_base_fee_params: BaseFeeParams,
+) -> Bytes {
+    let mut extra_data = [0u8; 9];
+    // If eip 1559 params are set, use them, otherwise use the canyon base fee param constants
+    // The eip 1559 params should exist here since there was a check previously
+    if attributes.eip_1559_params.unwrap() == B64::ZERO {
+        let mut default_params = [0u8; 8];
+        default_params[..4].copy_from_slice(
+            &(default_base_fee_params.max_change_denominator as u32).to_be_bytes(),
+        );
+        default_params[4..]
+            .copy_from_slice(&(default_base_fee_params.elasticity_multiplier as u32).to_be_bytes());
+        extra_data[1..].copy_from_slice(&default_params);
+        Bytes::copy_from_slice(&extra_data)
+    } else {
+        let (elasticity, denominator) =
+            decode_eip_1559_params(attributes.eip_1559_params.expect("eip_1559_params is none"));
+        let mut eip_1559_params = [0u8; 8];
+        println!("denominator: {}", denominator);
+        println!("elasticity: {}", elasticity);
+        eip_1559_params[..4].copy_from_slice(&(denominator as u32).to_be_bytes());
+        eip_1559_params[4..].copy_from_slice(&(elasticity as u32).to_be_bytes());
+        extra_data[1..].copy_from_slice(&eip_1559_params);
```

Comment:
would be nice to copy directly to the `extra_data` slice instead of using an intermediate value